### PR TITLE
Integration Test Driver: Improve documentation, Sync Docs with Code docstrings

### PR DESCRIPTION
**Pull request by tfc**
###### Description of changes

I gave a class this week and some participants correctly pointed out confusion with differences in the integration test driver's machine class doc strings and the nixos documentation.

This PR improves the situation by:

- moving the text from the nixos documentation to the python docstrings in the `Machine` class implementation
- auto-generate the nixos documentation markdown parts from the python docstrings
- improve some test script check output on the shell

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/lib/test-driver/test_driver/machine.py [hunk 1]
```diff
@@ -577,12 +629,13 @@ def shell_interact(self, address: Optional[str] = None) -> None:
...
 
     def console_interact(self) -> None:
-        """Allows you to interact with QEMU's stdin
-
-        The shell can be exited with Ctrl+D. Note that Ctrl+C is not allowed to be used.
-        QEMU's stdout is read line-wise.
-
-        Should only be used during test development, not in the production test."""
+        """
+        Allows you to directly interact with QEMU's stdin.
```
Comment (roberth) [inline nixos/lib/test-driver/test_driver/machine.py:633]:
```suggestion
        Allows you to directly interact with QEMU's stdin, by forwarding terminal input to the QEMU process.
```

File: nixos/lib/test-driver/test_driver/machine.py [hunk 2]
```diff
@@ -577,12 +629,13 @@ def shell_interact(self, address: Optional[str] = None) -> None:
...
+        This should only be used during test development, not in production
+        tests.
```
Comment (roberth) [inline nixos/lib/test-driver/test_driver/machine.py:635]:
```suggestion
        This is for use with the interactive test driver, not for production tests, which run unattended. 
```

File: nixos/lib/test-driver/test_driver/machine.py [hunk 3]
```diff
@@ -789,8 +875,16 @@ def copy_from_host_via_shell(self, source: str, target: str) -> None:
...
     def copy_from_host(self, source: str, target: str) -> None:
-        """Copy a file from the host into the guest via the `shared_dir` shared
-        among all the VMs (using a temporary directory).
+        """
+        Copies a file from host to machine, e.g.,
+        `copy_from_host("myfile", "/etc/my/important/file")`.
+
+        The first argument is the file on the host. The file needs to be
+        accessible while building the nix derivation. The second argument is
+        the location of the file on the machine.
```
Comment (roberth) [inline nixos/lib/test-driver/test_driver/machine.py:884]:
```suggestion
        The first argument is the file on the host. Note that the "host" refers
        to the environment in which the test driver runs, which is typically the
        Nix build sandbox.
        
        The second argument is the location of the file on the machine that will be written to.
```

Comment (roberth) [reply nixos/lib/test-driver/test_driver/machine.py:884]: I wonder about the permissions of the file.

File: nixos/lib/test-driver/test_driver/machine.py [hunk 4]
```diff
@@ -864,12 +987,9 @@ def screen_matches(last: bool) -> bool:
...
         """
-        Wait for the provided regex to appear on console.
-        For each reads,
-
-        If timeout is None, timeout is infinite.
-
-        `timeout` is in seconds.
+        Wait until the supplied regular expressions match a line of the
+        serial console output.
+        This method is useful when OCR is not possible or accurate enough.
```
Comment (roberth) [inline nixos/lib/test-driver/test_driver/machine.py:992]:
```suggestion
        This method is useful when OCR is not possible or inaccurate.
```

File: nixos/lib/testing/driver.nix [hunk 5]
```diff
@@ -79,6 +80,9 @@ let
 
         ${testDriver}/bin/generate-driver-symbols
         ${lib.optionalString (!config.skipLint) ''
+          echo "Linting test script (enable/disable: config.skipLint)"
+          echo "See https://nixos.org/manual/nixos/stable/#sec-test-options-reference"
```
Comment (roberth) [inline nixos/lib/testing/driver.nix:84]:
```suggestion
          echo "See https://nixos.org/manual/nixos/stable/#test-opt-skipLint"
```

File: nixos/lib/testing/driver.nix [hunk 6]
```diff
@@ -65,7 +65,8 @@ let
           echo "${builtins.toString vlanNames}" >> testScriptWithTypes
           echo -n "$testScript" >> testScriptWithTypes
 
-          cat -n testScriptWithTypes
+          echo "Running type check (enable/disable: config.skipTypeCheck)"
+          echo "See https://nixos.org/manual/nixos/stable/#sec-test-options-reference"
```
Comment (roberth) [inline nixos/lib/testing/driver.nix:69]:
```suggestion
          echo "See https://nixos.org/manual/nixos/stable/#test-opt-skipTypeCheck"
```

File: nixos/lib/test-driver/doc.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import ast
+import sys
+
+"""
+This program takes all the Machine class methods and prints its methods in
+markdown-style. These can then be included in the NixOS test driver
+documentation, which is then always in sync with the code.
```
Comment (roberth) [inline nixos/lib/test-driver/doc.py:7]:
```suggestion
markdown style, assuming the docstrings themselves are also in markdown.

These are included in the test driver documentation in the NixOS manual.
See https://nixos.org/manual/nixos/stable/#ssec-machine-objects
```

I think synchronicity is implied.

Comment (roberth) [reply nixos/lib/test-driver/doc.py:7]: I haven't thoroughly reviewed the python code below, but it looks like it gets the job done.
A bit ad hoc perhaps, but that's fine for the currently limited purpose.

**Review by roberth — COMMENTED** [review]
Awesome

File: nixos/lib/test-driver/doc.py (hunk 7 above)
Comment (tfc) [reply nixos/lib/test-driver/doc.py:7]: It's ad hoc, yes. The existing tools for docstring-to-markdown generation partly didn't like that we're using new things (e.g. the `|` sum type combinator) and also generated too much which would need to be filtered. so i figured that this approach here is even simpler and very easy to customize.

File: nixos/lib/test-driver/machine-docstrings.nix [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+, python3
+}:
+
+let
+  env = { nativeBuildInputs = [ python3 ]; };
+in
+
+runCommand "machine-docstrings.md" env ''
+  python3 ${./doc.py} ${./test_driver/machine.py} > $out
+''
```
[the remark this answers is not in the document]
Comment (tfc) [reply nixos/lib/test-driver/machine-docstrings.nix:11]: I incorporate all the things of this comment, but i don't understand how the renaming of things reduced the per derivation overhead - did i miss anything?

File: nixos/lib/test-driver/doc.py (hunk 7 above)
Comment (tfc) [reply nixos/lib/test-driver/doc.py]: Oh, i get it... you mean in the event of more stuff being extracted, having one deriv with multiple output files in the same output folder. yes.

File: nixos/lib/test-driver/machine-docstrings.nix (hunk 8 above)
[the remark this answers is not in the document]
Comment (roberth) [reply nixos/lib/test-driver/machine-docstrings.nix]: I was thinking into a future where we generate more files, in which case the previous code would suggest fine grained derivations. It was not a real problem yet. (The new names are an improvement today.)

File: nixos/doc/manual/default.nix [hunk 9]
```diff
@@ -63,6 +63,9 @@ let
       optionIdPrefix = "test-opt-";
     };
 
+  testDriverMachineDocstrings = pkgs.callPackage
+    "${prefix}/nixos/lib/test-driver/nixos-test-driver-docstrings.nix" {};
```
Comment (roberth) [inline nixos/doc/manual/default.nix:67]: OfBorg failed with

```
       … while realising the context of path '/nix/store/3jdj37ibg4nz1srpx6q9cl4siv2xam2a-ofborg-evaluator-0/nixos/lib/test-driver/nixos-test-driver-docstrings.nix'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-0/lib/customisation.nix:124:45:

          123|     let
          124|       f = if lib.isFunction fn then fn else import fn;
             |                                             ^
          125|       fargs = lib.functionArgs f;

       error: path '/nix/store/3jdj37ibg4nz1srpx6q9cl4siv2xam2a-ofborg-evaluator-0' is not valid
```

I think it's not happy about importing from a copied path, and indeed we shouldn't cause extra copying into the store.

This seems to work fine:

```suggestion
    ../../nixos/lib/test-driver/nixos-test-driver-docstrings.nix {};
```

**Review by roberth — APPROVED** [review]
Awesome!

**Pull request merged**